### PR TITLE
Make wrapIndexOnce check async, avoid DtoH sync on index_put_

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125952

Internal xref: https://fb.workplace.com/groups/1075192433118967/posts/1427156211255919/

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

**Review by lezcano — COMMENTED** [review]
We already do that within canDispatchToMaskedFill:

```python
static std::tuple<bool, Tensor> canDispatchToMaskedFill(const Tensor& self, const torch::List<c10::optional<at::Tensor>>& indices,
const Tensor& value){
  if (!(value.numel() ==1 && value.device().is_cpu())){
```

**Comment by ezyang** [comment]
OK, that's useful, because the internal user tested and they also said this did not fix it.

**Comment by ezyang** [comment]
I'm out of time right now, but here is the repro

```
# (c) Meta Platforms, Inc. and affiliates. Confidential and proprietary.

import logging
from dataclasses import dataclass
from datetime import datetime

import torch
import torch._inductor.config as inductor_config


logger = logging.getLogger(__name__)
TIME_FORMAT_STR: str = "%b_%d_%H_%M_%S"


@dataclass
class BenchmarkConfig:
    batch_size: int = 256
    enable_bf16: bool = True
    enable_pt2: bool = True
    device = "cuda:0"
    d_in = 2048


class SimpleModel(torch.nn.Module):
    def __init__(self, dim):
        super().__init__()
        self.linear1 = torch.nn.Linear(dim, dim, bias=False)
        self.ts_encoding_params_dict = torch.nn.Parameter(
            torch.empty(
                [
                    2000,
                    dim,
                ]
            ).uniform_(-0.01, 0.01)
        )
        self.linear2 = torch.nn.Linear(dim, dim, bias=False)

    def forward(
        self,
        x,
        num_object,
        user_event_ts_buckets,
    ):
        emb = self.linear1(x)
        # user_event_ts_encoding = self.ts_encoding_params_dict[user_event_ts_buckets, :]
        user_event_ts_encoding = self.ts_encoding_params_dict.index_select(
            0, user_event_ts_buckets
        )
        emb = emb + user_event_ts_encoding
        res = self.linear2(emb)
        return res


def create_model_input(benchmark_config: BenchmarkConfig):
    batch_size = benchmark_config.batch_size
    d_in = benchmark_config.d_in
    device = benchmark_config.device

    dtype = torch.bfloat16 if benchmark_config.enable_bf16 else torch.float32

    x = torch.rand(
        batch_size * 1000, d_in, dtype=dtype, device=device
    ).requires_grad_()  # assuming seq_len_per_example is max_length // 2
    num_object = torch.tensor(
        [1000] * batch_size,
        dtype=torch.int,
        device=device,
    )

    user_event_ts_buckets = torch.randint(
        0,
        2000,
        (1000 * batch_size,),
        dtype=torch.int,
        device=device,
    )

    return (
        x,
        num_object,
        user_event_ts_buckets,
    )


def run_first_model_once(model, input):
    pred = model(*input)
    pred[0].sum().backward()


def single_run_benchmark():
    benchmark_config = BenchmarkConfig()
    model_input = create_model_input(benchmark_config)
    model = SimpleModel(benchmark_config.d_in)

    if benchmark_config.enable_bf16:
        model = model.to(dtype=torch.bfloat16)

    if benchmark_config.enable_pt2:
        inductor_config.decompose_mem_bound_mm = True
        inductor_config.trace.enabled = True
        model = torch.compile(model)
        model = model.to(benchmark_config.device)
        print("Start compiling model.")
        run_first_model_once(model, model_input)
    else:
        model = model.to(benchmark_config.device)

    # trace
    with torch.profiler.profile(with_flops=True) as profiler:
        for _ in range(5):
            run_first_model_once(model, model_input)

    trace_file_prefix = "{}".format(
        datetime.now().strftime(TIME_FORMAT_STR),
    )
    profiler.export_chrome_trace(
        f"manifold://pyper_traces/tree/traces/test/{trace_file_prefix}_trace.json"
    )
    trace_url = f"https://our.internmc.facebook.com/intern/perfdoctor/trace_view?filepath=tree/traces/test/{trace_file_prefix}_trace.json.gz&bucket=pyper_traces"
    print(f"Trace link: {trace_url}")

    return


def main() -> None:
    single_run_benchmark()
    print("done")


if __name__ == "__main__":
    main()  # pragma: no cover
```

**Comment by lezcano** [comment]
I put up a fix, but I was not able to test whether it works (my triton version is acting up with the repro). Mind checking if it fixes the issue?

**Comment by ezyang** [comment]
OK, I got a better stacktrace here:

```
#12 at::_ops::item::call(at::Tensor const&) from ??:0                                                                            
#13 long at::Tensor::item<long>() const from ??:0                                                                                
#14 at::native::computeLinearIndex(at::Tensor const&, c10::ArrayRef<at::Tensor>, bool) [clone .isra.0] from tmpxft_00061ce6_00000
000-6_Indexing.cudafe1.cpp:0                                                                                                     
#15 at::native::makeLinearIndex(at::Tensor, c10::IListRef<at::OptionalTensorRef>, bool) [clone .constprop.0] from tmpxft_00061ce6
_00000000-6_Indexing.cudafe1.cpp:0                                                                                               
#16 at::native::(anonymous namespace)::index_put_with_sort_kernel(at::Tensor&, c10::List<std::optional<at::Tensor> > const&, at::
Tensor const&, bool, bool) from ??:0                                                                                             
#17 at::native::_index_put_impl_(at::Tensor&, c10::List<std::optional<at::Tensor> > const&, at::Tensor const&, bool, bool) from ?
?:0                
```

**Comment by ezyang** [comment]
It's this one:

```
static Tensor wrapIndexOnce(const Tensor & index, int64_t dim, int64_t dim_size, bool check_range=true) {
//we don't need to check range in backward - if there were out of bounds indices forward should already have errored out
  if (index.numel() != 0 && check_range) {
    TORCH_INTERNAL_ASSERT(0);
    auto max_idx = index.max().item<int64_t>();
    auto min_idx = index.min().item<int64_t>();
    if (max_idx >= dim_size) {
      TORCH_CHECK_INDEX(false, "index ", max_idx, " is out of bounds for dimension ", dim, " with size ", dim_size);
    } 
    if (min_idx < -dim_size) {
      TORCH_CHECK_INDEX(false, "index ", min_idx, " is out of bounds for dimension ", dim, " with size ", dim_size);
    } 
  }
  return index.remainder(dim_size);
} 
```

and the reason it doesn't sync in eager is because, dun dun dun, eager used some special API to avoid the check range.

**Comment by ezyang** [comment]
Updated with fix that actually is confirmed to work

File: aten/src/ATen/native/cuda/Indexing.cu [hunk 1]
```diff
@@ -334,14 +335,8 @@ static ReduceMaximum reduce_maximum;
...
-    auto max_idx = index.max().item<int64_t>();
-    auto min_idx = index.min().item<int64_t>();
-    if (max_idx >= dim_size) {
-      TORCH_CHECK_INDEX(false, "index ", max_idx, " is out of bounds for dimension ", dim, " with size ", dim_size);
-    }
-    if (min_idx < -dim_size) {
-      TORCH_CHECK_INDEX(false, "index ", min_idx, " is out of bounds for dimension ", dim, " with size ", dim_size);
-    }
+    at::_assert_async(index.max() < dim_size);
+    at::_assert_async(index.min() >= -dim_size);
```
Comment (lezcano) [inline aten/src/ATen/native/cuda/Indexing.cu:339]: I didn't know we had these working in the end. How does compile interprete these?

**Review by lezcano — APPROVED** [review]
Sounds good if it fixes the problem. Could you add a regression test?

File: aten/src/ATen/native/cuda/Indexing.cu (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/native/cuda/Indexing.cu:339]: This is the CUDA kernel, so compile doesn't really interact with this in a nontrivial way, you just end up hitting this at runtime.

**Pull request closed without merging**